sns_module/src/main.rs

```diff
@@ -0,0 +1,211 @@
+    replace_state(state);
+}
+
+#[post_upgrade]
+fn post_upgrade() {
+    let state = sns_module::memory::get_state().unwrap();
+    replace_state(state);
+}
+
+#[query]
+fn get_status() -> Status {
+    let balances = get_principal_to_icp();
+    read_state(|s| Status {
+        participants: balances.len(),
+        total_icp_deposited: balances.iter().map(|(_, b)| b).sum(),
+        time_left: s.end_ts.saturating_sub(ic_cdk::api::time()),
+        start_at: s.start_ts,
+        end_at: s.end_ts,
+        minimum_deposit_amount: MIN_DEPOSIT_AMOUNT,
+    })
+}
+
+#[query]
+fn get_principal_to_icp_deposited() -> Vec<(Principal, u64)> {
+    sns_module::memory::get_principal_to_icp()
+}
+
+#[query]
+fn get_icp_deposited(of: Principal) -> u64 {
+    sns_module::memory::get_icp_deposited(of)
+}
+
+#[query]
+fn get_principal_to_wtn_allocation() -> Vec<(Principal, u64)> {
+    sns_module::memory::get_principal_to_wtn_owed()
+}
+
+#[query]
+fn get_wtn_allocated(of: Principal) -> u64 {
+    sns_module::memory::get_wtn_owed(of)
+}
+
+#[update]
+fn get_icp_deposit_address(target: Principal) -> AccountIdentifier {
+    let subaccount = Subaccount(derive_staking(target));
+    let self_id: PrincipalId = ic_cdk::id().into();
+    AccountIdentifier::new(self_id, Some(subaccount))
+}
+
+#[update]
+async fn notify_icp_deposit(target: Principal, amount: u64) -> Result<u64, String> {
+    if amount < MIN_DEPOSIT_AMOUNT {
+        return Err(format!(
+            "Amount lower than the minimum deposit: {} ICP",
+            MIN_DEPOSIT_AMOUNT / E8S
+        ));
+    }
+    is_swap_available()?;
+    let icp_ledger = read_state(|s| s.icp_ledger_id);
+    let received_tokens = amount.checked_sub(10_000).unwrap();
+    match transfer(
+        Some(derive_staking(target)),
+        ic_cdk::id(),
+        Nat::from(received_tokens),
+        None,
+        icp_ledger,
+    )
+    .await
+    {
+        Ok(block_index) => {
+            deposit_icp(target, received_tokens);
+            Ok(block_index)
+        }
+        Err(e) => Err(format!("{e}")),
+    }
+}
+
+#[update]
+async fn distribute_tokens() -> Result<(), String> {
+    assert_eq!(
+        ic_cdk::caller(),
+        Principal::from_text("bo5bf-eaaaa-aaaam-abtza-cai").unwrap()
+    );
+    if !get_principal_to_wtn_owed().is_empty() {
+        return Err("Already distributed WTN".to_string());
+    }
+    if !is_distribution_available() {
+        return Err("Distribution not available".to_string());
+    }
+
+    let (icp_ledger, wtn_ledger) = read_state(|s| (s.icp_ledger_id, s.wtn_ledger_id));
+    let icp_balance = balance_of(ic_cdk::id(), icp_ledger).await?;
+    let wtn_balance = balance_of(ic_cdk::id(), wtn_ledger).await?;
+    if wtn_balance < 100_000_000 {
+        return Err("Nothing to distribute".to_string());
+    }
+
+    let balances = sns_module::memory::get_principal_to_icp();
+    let total_tracked_icp: u64 = balances.iter().map(|(_, b)| b).sum();
```

Comment:
you compute the `total_tracked_icp` here and call `dispatch_tokens` afterwards which re-compute `total_tracked`. By passing the total_tracked as a parameter of the function you can make it O(n) instead of O(2n)